apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerConnection.cs

```diff
@@ -446,10 +464,16 @@ internal void ReleaseSession(Session session)
 
         private Task<Session> AllocateSession(TransactionOptions options, CancellationToken cancellationToken)
         {
-            return ExecuteHelper.WithErrorTranslationAndProfiling(() =>
+            return ExecuteHelper.WithErrorTranslationAndProfiling(async () =>
             {
                 AssertOpen("execute command");
                 Task<Session> result;
+                // If the shared session gets used for anything but a write, we need
+                // to clear out the transaction state, because this will happen implicitly
+                // by spanner when a read is done without the active transaction.
+                // For other cases, the transaction state is always cleared as soon as its
+                // placed back into the session pool.
+                bool sharedSessionReadOnlyUse = false;
```

Comment:
Suggested optimization from spanner is to "prewarm" a session with a readwrite transaction.  This code could be a little tighter.  I might try to refactor it so that you always get a session along with the transaction in a tuple and there is less state that has to be managed -- because part of the complicated bit here is knowing when to invalidate the cache or not.
It would be super helpful if I could modify the partial Session class to add some of my own state.  The fact that I cannot add state directly to the class creates some unnecessary complications.